e2e/services/browser.ts

```
@@ -0,0 +1,92 @@
+
+type ConnectRetryParams = { url?: string; timeout?: number; interval?: number };
+
+const sleep = ms => new Promise(resolve => setTimeout(resolve, ms));
+
+export class TestBrowserService {
+  static #remoteBrowser: Browser;
+
+  static async getRemoteBrowser(
+    {
+      url = "http://localhost:61338",
+      timeout = 3000,
+      interval = 250,
+    }: ConnectRetryParams = {
+      url: "http://localhost:61338",
+      timeout: 3000,
+      interval: 250,
+    }
+  ) {
+    if (TestBrowserService.#remoteBrowser) {
+      return TestBrowserService.#remoteBrowser;
+    }
+
+    const startTime = Date.now();
+    async function connectLoop() {
+      try {
+        const remoteChromium = await chromium.connectOverCDP(url);
+        TestBrowserService.#remoteBrowser = remoteChromium;
+        return remoteChromium;
+      } catch (e) {
+        const elapsedTime = Date.now() - startTime;
+        if (elapsedTime > timeout)
+          throw new Error(`Can't connect to a browser at ${url}`);
+        await sleep(interval);
+        return connectLoop();
+      }
+    }
+
+    return connectLoop();
+  }
+
+  static async getRemoteBrowserContext(connectParams?: ConnectRetryParams) {
+    const browserContext = (
+      await TestBrowserService.getRemoteBrowser(connectParams)
+    ).contexts()[0];
+    if (!browserContext)
+      throw new Error("Not connected to a remote browser with contexts.");
+    return browserContext;
+  }
+
+  static async getRemoteBrowserPage(connectParams?: ConnectRetryParams) {
+    const browserContext = await TestBrowserService.getRemoteBrowserContext(
+      connectParams
+    );
+
+    while (browserContext.pages().length === 0 || !browserContext.pages()[0]) {
+      await sleep(100);
+    }
+
+    return browserContext.pages()[0];
+  }
+
+  static async closeRemoteBrowser(connectParams?: ConnectRetryParams) {
+    const browser = await TestBrowserService.getRemoteBrowser(connectParams);
+    await browser.close();
```

would add an empty catch here, just in case.